libfwupdplugin/fu-bluez-device.c

```diff
@@ -293,59 +291,151 @@ fu_bluez_device_get_ble_string_property(const gchar *obj_path,
+{
+	g_autoptr(GError) error_local = NULL;
+	g_autofree gchar *obj_uuid =
+	    fu_bluez_device_get_interface_uuid(self, obj, obj_path, iface_name);
+
+	if (obj_uuid == NULL)
+		return FALSE;
+
+	/* Register device by service UUID. */
+	fu_device_add_instance_str(FU_DEVICE(self), "GATT", obj_uuid);
+	if (!fu_device_build_instance_id_full(FU_DEVICE(self),
+					      FU_DEVICE_INSTANCE_FLAG_VISIBLE |
+						  FU_DEVICE_INSTANCE_FLAG_QUIRKS,
+					      &error_local,
+					      "BLUETOOTH",
+					      "GATT",
+					      NULL)) {
+		g_debug("failed to register %s service: %s", obj_uuid, error_local->message);
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_bluez_device_read_battery_interface(FuBluezDevice *self,
+				       GDBusObject *obj,
+				       const gchar *obj_path,
+				       const gchar *iface_name)
+{
+	guint8 percentage = FWUPD_BATTERY_LEVEL_INVALID;
+	g_autoptr(GDBusInterface) iface = NULL;
+	g_autoptr(GError) error_local = NULL;
+	g_autoptr(GVariant) obj_percentage = NULL;
+
+	iface = g_dbus_object_get_interface(obj, iface_name);
+	if (iface == NULL)
+		return FALSE;
+
+	obj_percentage =
+	    fu_bluez_device_get_ble_property(obj_path, iface_name, "Percentage", &error_local);
+	if (obj_percentage == NULL) {
+		/* Sometimes battery service announced but has no value, no error in that case. */
+		g_debug("failed to get battery percentage from org.bluez.Battery1");
+		return FALSE;
+	}
+
+	percentage = g_variant_get_byte(obj_percentage);
+
+	fu_device_set_battery_level(FU_DEVICE(self), percentage);
+
+	return TRUE;
+}
+
 /*
  * Populates the {uuid_helper : object_path} entries of a device for all its
  * characteristics.
- *
- * TODO: Extend to services and descriptors too?
  */
 static void
-fu_bluez_device_add_device_uuids(FuBluezDevice *self)
+fu_bluez_device_ensure_gatt_interfaces(FuBluezDevice *self)
 {
 	FuBluezDevicePrivate *priv = GET_PRIVATE(self);
 	g_autolist(GDBusObject) obj_list = NULL;
 
 	obj_list = g_dbus_object_manager_get_objects(priv->object_manager);
 	for (GList *l = obj_list; l != NULL; l = l->next) {
 		GDBusObject *obj = G_DBUS_OBJECT(l->data);
-		g_autoptr(GDBusInterface) iface = NULL;
-		g_autoptr(GError) error_local = NULL;
-		g_autofree gchar *obj_uuid = NULL;
 		const gchar *obj_path = g_dbus_object_get_object_path(obj);
 
 		/* not us */
 		if (!g_str_has_prefix(g_dbus_object_get_object_path(obj),
 				      g_dbus_proxy_get_object_path(priv->proxy)))
 			continue;
 
-		/*
-		 * TODO: Currently restricted to getting UUIDs for
-		 * characteristics only, as the only use case we're
-		 * going to need for now is reading/writing
-		 * characteristics.
-		 */
-		iface = g_dbus_object_get_interface(obj, "org.bluez.GattCharacteristic1");
-		if (iface == NULL)
+		/* Add characteristics UUID for reading/writing. */
+		if (fu_bluez_device_add_characteristic_uuid(self,
+							    obj,
+							    obj_path,
+							    "org.bluez.GattCharacteristic1"))
+			continue;
```

Comment:
I don't think so to be honest. Since characteristics are always part of a service:
_Services, Characteristics and Descriptors are organised in a hierarchy with Services at the top and Descriptors at the bottom. Services contain one or more Characteristics. A Characteristic owns zero or more Descriptors because Descriptors are completely optional whereas a Service must contain at least one Characteristic._ (c) https://www.bluetooth.com/blog/a-developers-guide-to-bluetooth/

Reply:
Might we not want to match on either? Don't worry about adding dozens of quirk-only GUIDs, it's 100% okay.

Reply:
I've refactored things a bit; does my fixup still work?